File: VC_HIL/python/inverterhil/sensor_bus_mux.py
import math


def _normalize(source, kind):
    blank = {"id": 0, "dlc": 0, "payload": [0] * 8, "timestampS": math.nan, "sequence": 0, "valid": False, "kind": kind}
    if not isinstance(source, dict) or not source.get("valid"): return blank
    for name in ("id", "dlc", "payload", "timestampS", "sequence"):
        if name not in source: raise ValueError(f"Sensor frame lacks {name}.")
    blank.update(id=int(source["id"]), dlc=int(source["dlc"]), payload=[int(x) & 255 for x in source["payload"]],
                 timestampS=float(source["timestampS"]), sequence=int(source["sequence"]), valid=bool(source["valid"]))
    return blank


def sensorBusMux(imuFrame, steeringFrame):
    return sorted((x for x in (_normalize(imuFrame, "imu"), _normalize(steeringFrame, "steering")) if x["valid"]),
                  key=lambda x: (x["timestampS"], x["sequence"]))

File: VC_HIL/inverter/imu/sensorBusMux.m
function frames = sensorBusMux(imuFrame, steeringFrame)
%SENSORBUSMUX Combine due sensor frames without cross-channel overwriting.
frames = repmat(emptyFrame(), 1, 2);
frames(1) = normalizeFrame(imuFrame, 'imu');
frames(2) = normalizeFrame(steeringFrame, 'steering');
valid = [frames.valid];
frames = frames(valid);
if numel(frames) > 1
    timestamps = [frames.timestampS];
    sequences = [frames.sequence];
    [~, order] = sortrows([timestamps(:), double(sequences(:))], [1 2]);
    frames = frames(order);
end
end

function frame = normalizeFrame(source, kind)
frame = emptyFrame();
frame.kind = kind;
if isempty(source) || ~isstruct(source) || ~isfield(source, 'valid') || ...
        ~source.valid
    return;
end
required = {'id', 'dlc', 'payload', 'timestampS', 'sequence'};
for index = 1:numel(required)
    if ~isfield(source, required{index})
        error('inverterhil:MalformedSensorFrame', ...
            'Sensor frame lacks %s.', required{index});
    end
end
frame.id = uint32(source.id);
frame.dlc = uint8(source.dlc);
frame.payload = uint8(source.payload(:).');
frame.timestampS = double(source.timestampS);
frame.sequence = uint32(source.sequence);
frame.valid = logical(source.valid);
frame.kind = kind;
end

function frame = emptyFrame()
frame = struct('id', uint32(0), 'dlc', uint8(0), ...
    'payload', zeros(1, 8, 'uint8'), 'timestampS', NaN, ...
    'sequence', uint32(0), 'valid', false, 'kind', '');
end
